### PR TITLE
Revision and cleanup of Parser.py file, and fixed printing bug (spaces)

### PR DESCRIPTION
**Pull request by kommisar5150**
Did a bit of cleanup in the parser file, added descriptions for less evident functions. Also fixed a bug where .dataAlpha strings would not print multiple spaces (the line was already split, thus removing spaces).

### CODE REVIEW

File: ToolChain/Assembler/Parser/Parser.py [hunk 1]
```diff
@@ -116,37 +124,46 @@ def _parseCodeFile(self):
...
             lineNo += 1
+
             try:
                 sline = line.split()
                 if len(sline) > 0:
-                    # We are here, we have a code line
-                    buildInstruction, lineInstruction = self._parseCodeLine(sline)
+                    # If we make it here, there was something on the line
+                    # We move to the next step and parse the line to get the appropriate code for what's written
+
```
Comment (cannotparse) [inline ToolChain/Assembler/Parser/Parser.py]: Extra whitespace

File: ToolChain/Assembler/Parser/Parser.py [hunk 2]
```diff
@@ -116,37 +124,46 @@ def _parseCodeFile(self):
...
-        relativeAddressCounter = 0
-        lineNo = 0
+
+        instructionList = []            # Will contain the list of binary instructions
+        memoryReferenceDict = {}        # Contains memory references. Will contain label along with its relative address
+        relativeAddressCounter = 0      # Counter cumulatively adds length of instructions.
+        lineNo = 0                      # Keeps track of current line being evaluated to print errors at specific lines
 
         for line in self.fileContent:
+
```
Comment (cannotparse) [inline ToolChain/Assembler/Parser/Parser.py]: Extra whitespace

File: ToolChain/Assembler/Parser/Parser.py [hunk 3]
```diff
@@ -162,239 +179,128 @@ def _parseCodeLine(self, line: list=None):
...
         finalCode = 0x00
         form = None
         possibleCodes = None
 
+        # We evaluate the line and extract any bits of relevant info to build the appropriate instruction
         self._evaluateAndExtractInfoFromLine(line, buildInstruction)
 
         if type(buildInstruction.instructionCode) is not str:
-            # Instruction is real not just memory address or comment
+
```
Comment (cannotparse) [inline ToolChain/Assembler/Parser/Parser.py]: ws

File: ToolChain/Assembler/Parser/Parser.py [hunk 4]
```diff
@@ -162,239 +179,128 @@ def _parseCodeLine(self, line: list=None):
...
+            # We assume it's an instruction (integer), since memory references or comments would be strings
+            # Parser will verify all possible codes, and generate the appropriate instruction based on its form
             possibleCodes = self._findPossibleCodesForInstruction(buildInstruction)
             instructionCode, instructionForm = self._getInstructionCodeAndFormUsingPossibleCodes(buildInstruction,
                                                                                                  possibleCodes)
+
```
Comment (cannotparse) [inline ToolChain/Assembler/Parser/Parser.py]: ws

File: ToolChain/Assembler/Parser/Parser.py [hunk 5]
```diff
@@ -162,239 +179,128 @@ def _parseCodeLine(self, line: list=None):
...
             if instructionForm is None or instructionCode is None:
+                # Code evaluated in function isn't valid
                 raise ValueError("Assembler general error")
```
Comment (cannotparse) [inline ToolChain/Assembler/Parser/Parser.py]: Comment, code, and error message don't match. 
If anything, the line of code couldn't be recognized. Not sure what edge case triggers this. Would be nice to verify if this will cause it to print the line from the input file. I know some do.

File: ToolChain/Assembler/Parser/Parser.py [hunk 6]
```diff
@@ -162,239 +179,128 @@ def _parseCodeLine(self, line: list=None):
...
             else:
+                # Code is valid, fetch the instruction's form based on the arguments, then generate the instruction
                 buildInstruction.instructionCode = instructionCode
                 buildInstruction.instructionLength = formDescription[instructionForm]["length"]
                 lineInstruction = self._generateBinaryInstructionFromInstruction(buildInstruction,
                                                                                  formDescription[instructionForm])
+
         else:
+            # In this case, there were no instructions. Thus, we look at other possibilities
+
```
Comment (cannotparse) [inline ToolChain/Assembler/Parser/Parser.py]: ws

File: ToolChain/Assembler/Parser/Parser.py [hunk 7]
```diff
@@ -162,239 +179,128 @@ def _parseCodeLine(self, line: list=None):
...
             if COMMENT_INDICATORS in buildInstruction.instructionCode:
                 # Simply ignore a comment line
                 buildInstruction = None
                 lineInstruction = None
+
+            #  In these next cases, we can treat them as "data" variables with no real instructions.
+            #  We store the data as the instruction's source immediate value, and set instruction code to 0
+
```
Comment (cannotparse) [inline ToolChain/Assembler/Parser/Parser.py]: ws

File: ToolChain/Assembler/Parser/Parser.py [hunk 8]
```diff
@@ -162,239 +179,128 @@ def _parseCodeLine(self, line: list=None):
...
                 lineInstruction = buildInstruction.sourceImmediate
                 buildInstruction.instructionCode = 0
+
             else:
+                # Having exhausted all options, we assume there was nothing else on this line.
                 lineInstruction = None
 
         return buildInstruction, lineInstruction
 
-    def translateRegisterNameToRegisterCode(self, registerName: str=""):
```
Comment (cannotparse) [inline ToolChain/Assembler/Parser/Parser.py]: Why did you move this?

File: ToolChain/Assembler/Parser/Parser.py [hunk 9]
```diff
@@ -162,239 +179,128 @@ def _parseCodeLine(self, line: list=None):
...
-        This is an helper method. The aim to this is to keep caller code cleaner
-        TODO: refactor this... it's ugly
+        This is a helper method. The aim to this is to keep caller code cleaner
+        Method evaluates each part of the line to determine whether we're dealing with an instruction, label, data, etc.
+        We begin by verifying if the first argument has label or data mnemonics (not an instruction).
+        If none are found, we assume it's an instruction. We then proceed to evaluating the arguments.
+        The method uses the Instruction class' source and destination fields to enforce proper instruction format.
+        For example, if the source and destination have already been found, additional arguments means the line is an
+        invalid instruction.
+
```
Comment (cannotparse) [inline ToolChain/Assembler/Parser/Parser.py]: ws

File: ToolChain/Assembler/Parser/Parser.py [hunk 10]
```diff
@@ -162,239 +179,128 @@ def _parseCodeLine(self, line: list=None):
...
+                          MEMORY_REFERENCE_INDICATORS,
+                          DATA_ALPHA_INDICATOR,
+                          DATA_NUMERIC_INDICATOR,
+                          DATA_MEMORY_REFERENCE,
+                          EXPORTED_REFERENCE_INDICATOR]
+
+        sline = line.split()
+        buildInstruction.operationMnemonic = sline[0].upper()
+
+        # PART 1: Evaluate the first part of the line, which should be an instruction or mnemonic
```
Comment (cannotparse) [inline ToolChain/Assembler/Parser/Parser.py]: Actually, the first part looks for items that aren't instructions such as labels (mem refs) and handles them differently.

File: ToolChain/Assembler/Parser/Parser.py [hunk 11]
```diff
@@ -116,37 +124,46 @@ def _parseCodeFile(self):
...
+
             try:
                 sline = line.split()
                 if len(sline) > 0:
-                    # We are here, we have a code line
-                    buildInstruction, lineInstruction = self._parseCodeLine(sline)
+                    # If we make it here, there was something on the line
+                    # We move to the next step and parse the line to get the appropriate code for what's written
+
+                    buildInstruction, lineInstruction = self._parseCodeLine(line)
```
Comment (cannotparse) [inline ToolChain/Assembler/Parser/Parser.py]: Why change this to sending a line?

By sending a line, you are forcing each function to re-split the line. Not that it's a crazy expensive operation but it doesn't make sense to split it at line 140, verify the length, then discard the result to regenerate it later.

File: ToolChain/Assembler/Parser/Parser.py [hunk 12]
```diff
@@ -116,37 +124,46 @@ def _parseCodeFile(self):
...
             except Exception as e:
+                # If all else fails, the line has invalid code
                 print("ERROR while asssembling line {}: '{}'".format(str(lineNo), line,))
                 print(e)
                 quit()
 
         return instructionList, memoryReferenceDict, relativeAddressCounter
 
-    def _parseCodeLine(self, line: list=None):
+    def _parseCodeLine(self, line: str=None):
```
Comment (cannotparse) [inline ToolChain/Assembler/Parser/Parser.py]: See comment on splitting

File: ToolChain/Assembler/Parser/Parser.py [hunk 13]
```diff
@@ -162,239 +179,128 @@ def _parseCodeLine(self, line: list=None):
...
+        foundSource = False                              # Flag for whether we found a source immediate or register
+        foundDestination = False                         # Flag for whether we found a destination immediate or register
+        indicator_list = [COMMENT_INDICATORS,            # List of non-instruction indicators
+                          MEMORY_REFERENCE_INDICATORS,
+                          DATA_ALPHA_INDICATOR,
+                          DATA_NUMERIC_INDICATOR,
+                          DATA_MEMORY_REFERENCE,
+                          EXPORTED_REFERENCE_INDICATOR]
+
+        sline = line.split()
```
Comment (cannotparse) [inline ToolChain/Assembler/Parser/Parser.py]: See comment about splitting above

File: ToolChain/Assembler/Parser/Parser.py [hunk 14]
```diff
@@ -517,6 +429,68 @@ def _getInstructionCodeAndFormUsingPossibleCodes(self,
...
+            -Local memory reference label
+            -Comment
+            -.dataAlpha, .dataNumeric, or .dataMemref
+
+        :param line:
+        :param buildInstruction:
+        :return:
+        '''
+
+        sline = line.split()
```
Comment (cannotparse) [inline ToolChain/Assembler/Parser/Parser.py]: See line splitting.
This is the third split deep.

File: ToolChain/Assembler/Parser/Parser.py [hunk 15]
```diff
@@ -162,239 +179,128 @@ def _parseCodeLine(self, line: list=None):
...
+                          DATA_ALPHA_INDICATOR,
+                          DATA_NUMERIC_INDICATOR,
+                          DATA_MEMORY_REFERENCE,
+                          EXPORTED_REFERENCE_INDICATOR]
+
+        sline = line.split()
+        buildInstruction.operationMnemonic = sline[0].upper()
+
+        # PART 1: Evaluate the first part of the line, which should be an instruction or mnemonic
+
```
Comment (cannotparse) [inline ToolChain/Assembler/Parser/Parser.py]: ws

File: ToolChain/Assembler/Parser/Parser.py [hunk 16]
```diff
@@ -162,239 +179,128 @@ def _parseCodeLine(self, line: list=None):
...
+        if (MEMORY_REFERENCE_INDICATORS == sline[0][-1]) or COMMENT_INDICATORS == sline[0][0]:
```
Comment (cannotparse) [inline ToolChain/Assembler/Parser/Parser.py]: isn't `_parseindicatordata` responsible for this?

File: ToolChain/Assembler/Parser/Parser.py [hunk 17]
```diff
@@ -162,239 +179,128 @@ def _parseCodeLine(self, line: list=None):
...
-        if DATA_MEMORY_REFERENCE in buildInstruction.operationMnemonic:
-            # This is a memory reference used as a constant that needs to be
-            # linked later in the process.
-            if line[1][0] == ":":
-                line[1] = line[1][1:]
-            buildInstruction.sourceImmediate = (":" + line[1] + ":").upper().encode("utf-8")
-            buildInstruction.instructionLength = 4
-            buildInstruction.instructionCode = DATA_MEMORY_REFERENCE
+
+        if buildInstruction.operationMnemonic in indicator_list:
```
Comment (cannotparse) [inline ToolChain/Assembler/Parser/Parser.py]: I mean, some parts of this I like but overall it still uses these indicators to skip ahead rather than left-right parsing.

I'd strongly recommend someone implement an LR0 for this. Would be a good first commit.

File: ToolChain/Assembler/Parser/Parser.py [hunk 18]
```diff
@@ -162,239 +179,128 @@ def _parseCodeLine(self, line: list=None):
...
+            # If the first argument of the line isn't an instruction, memory reference, or comment,
+            # it could be an indicator
+            self._parseIndicatorData(line, buildInstruction)
             return
 
-        # We don't want to redo the first part. That would cause problem with in code memory references.
-        for part in line[1:]:
+        # PART 2: Evaluate arguments
```
Comment (cannotparse) [inline ToolChain/Assembler/Parser/Parser.py]: Inconsistent comment style and indentation. Previously, comments would be ahead of the section being described.

File: ToolChain/Assembler/Parser/Parser.py [hunk 19]
```diff
@@ -162,239 +179,128 @@ def _parseCodeLine(self, line: list=None):
...
+
```
Comment (cannotparse) [inline ToolChain/Assembler/Parser/Parser.py]: ws

File: ToolChain/Assembler/Parser/Parser.py [hunk 20]
```diff
@@ -162,239 +179,128 @@ def _parseCodeLine(self, line: list=None):
...
+        # We continue evaluating the line to find the possible arguments of an instruction since no mnemonics were found
+        # First part of the line would theoretically be the instruction, so we omit it.
+        # We keep track of whether or not we found the source and destination of a particular instruction.
+        # This allows us to determine if the instruction has extra arguments, rendering it invalid.
+        # Handling for single argument instructions is handled in the findPossibleCodesForInstruction method.
+
+        for part in sline[1:]:
+
```
Comment (cannotparse) [inline ToolChain/Assembler/Parser/Parser.py]: ws

File: ToolChain/Assembler/Parser/Parser.py [hunk 21]
```diff
@@ -440,19 +348,21 @@ def _evaluateAndExtractInfoFromLine(self, line, buildInstruction):
                     part = part[1:]
                 if not foundSource:
                     foundSource = True
-                    buildInstruction.sourceImmediate = MEMORY_REFERENCE_INDICATORS+ part.upper() + MEMORY_REFERENCE_INDICATORS
+                    buildInstruction.sourceImmediate = ":" + part.upper() + ":"
```
Comment (cannotparse) [inline ToolChain/Assembler/Parser/Parser.py]: Leaving the long form lets us know why specifically `:` is used. While I like it short and sweet, the rest of the code uses the long form and it keeps things orderly.

File: ToolChain/Assembler/Parser/Parser.py [hunk 22]
```diff
@@ -440,19 +348,21 @@ def _evaluateAndExtractInfoFromLine(self, line, buildInstruction):
...
                 else:
                     if foundDestination:
                         raise ValueError("Invalid operation format")
                     foundDestination = True
-                    buildInstruction.destinationImmediate = MEMORY_REFERENCE_INDICATORS + part.upper() + MEMORY_REFERENCE_INDICATORS
+                    buildInstruction.destinationImmediate = ":" + part.upper() + ":"
```
Comment (cannotparse) [inline ToolChain/Assembler/Parser/Parser.py]: keep long form

File: ToolChain/Assembler/Parser/Parser.py [hunk 23]
```diff
@@ -440,19 +348,21 @@ def _evaluateAndExtractInfoFromLine(self, line, buildInstruction):
...
+
                 continue
 
     def _findPossibleCodesForInstruction(self, partialInstruction: Instruction=None):
         """
         This will find the possible instruction codes for a given partial instruction.
         It will return the complete list of possible instruction code. Another method will be used
         to sort out the bad codes.
+
```
Comment (cannotparse) [inline ToolChain/Assembler/Parser/Parser.py]: ws

File: ToolChain/Assembler/Parser/Parser.py [hunk 24]
```diff
@@ -517,6 +429,68 @@ def _getInstructionCodeAndFormUsingPossibleCodes(self,
 
         return instructionCode, instructionForm
 
+    def _parseIndicatorData(self, line, buildInstruction):
+
```
Comment (cannotparse) [inline ToolChain/Assembler/Parser/Parser.py]: ws

File: ToolChain/Assembler/Parser/Parser.py [hunk 25]
```diff
@@ -508,7 +419,8 @@ def _getInstructionCodeAndFormUsingPossibleCodes(self,
                                  "width",
                                  "flags",
                                  "instructionLength"]
-        # This deals with overlong instructions ex: JMP $A $B <- one too many register
+
```
Comment (cannotparse) [inline ToolChain/Assembler/Parser/Parser.py]: ws

File: ToolChain/Assembler/Parser/Parser.py [hunk 26]
```diff
@@ -538,18 +512,17 @@ def _generateBinaryInstructionFromInstruction(self, instruction: Instruction=Non
...
             while instructionPartBitMask % 2 is not 1:
                 instructionPartBitMask >>= 1  # Shift right one bit
                 shiftValue += 1
 
+            # We grab the value of the instruction to determine whether it's an actual instruction or a memory reference
             instructionPartOriginal = getattr(instruction, instructionPart)
-            # We need an intermediate step because a part could be text...
+            # We need an intermediate step because a part could be text
+
```
Comment (cannotparse) [inline ToolChain/Assembler/Parser/Parser.py]: ws

File: ToolChain/Assembler/Parser/Parser.py [hunk 27]
```diff
@@ -517,6 +429,68 @@ def _getInstructionCodeAndFormUsingPossibleCodes(self,
 
         return instructionCode, instructionForm
 
+    def _parseIndicatorData(self, line, buildInstruction):
+
+        '''
```
Comment (cannotparse) [inline ToolChain/Assembler/Parser/Parser.py]: use `"""` to match the rest and most other python 3 docstrings

File: ToolChain/Assembler/Parser/Parser.py [hunk 28]
```diff
@@ -517,6 +429,68 @@ def _getInstructionCodeAndFormUsingPossibleCodes(self,
...
+        :return:
+        '''
+
+        sline = line.split()
+
+        if EXPORTED_REFERENCE_INDICATOR in buildInstruction.operationMnemonic:
+            # Global memory reference which can be used by other files. This is a special case
+            buildInstruction.instructionCode = sline[0].upper() + " "
+            buildInstruction.instructionCode += sline[1].replace(MEMORY_REFERENCE_INDICATORS, "").upper()
+
```
Comment (cannotparse) [inline ToolChain/Assembler/Parser/Parser.py]: ws

File: ToolChain/Assembler/Parser/Parser.py (hunk 8 above)
Comment (cannotparse) [reply ToolChain/Assembler/Parser/Parser.py]: Yeah, provided an explanation for moving it. Could be alphabetical or order of appearance, but make sure everything else matches that scheme.

File: ToolChain/Assembler/Parser/Parser.py [hunk 29]
```diff
@@ -517,6 +429,68 @@ def _getInstructionCodeAndFormUsingPossibleCodes(self,
...
+        if DATA_ALPHA_INDICATOR in buildInstruction.operationMnemonic:
+            # Treat this as an alphabetic field
+            # The alpha value will be put in the source immediate value as well
+            # We need to rebuild the alpha into a byte string to be read by the VM
+            # We want to preserve spaces in the original string, so we limit the split to the first argument
+
+            textline = line.split(maxsplit=1)
+            sline = line.split()
+            alpha = b"" + textline[1].encode("utf-8") + b" "
```
Comment (cannotparse) [inline ToolChain/Assembler/Parser/Parser.py]: I don't think the extra splitting of the line is worth this simplification. While I think this is nice, it doesn't really add much benefit. 

If you want to simplify this, using the original fragment:
```python3
alpha = b""	
alpha += [part.encode("utf-8") + b" " for part in line[1:]]	
alpha = alpha[:-1] + b"\x00"  # Remove trailing space	 Strings are null terminated	
buildInstruction.sourceImmediate = alpha	
buildInstruction.instructionLength = len(alpha)	
buildInstruction.instructionCode = line[0].upper()
```

But I'm weary even of my own suggestion as needless wizardry. There are plenty of zip and joins that can accomplish this as well.

File: ToolChain/Assembler/Parser/Parser.py [hunk 30]
```diff
@@ -573,3 +546,128 @@ def _generateBinaryInstructionFromInstruction(self, instruction: Instruction=Non
...
         return binaryInstructionByteString
+
+    def translateRegisterNameToRegisterCode(self, registerName: str=""):
+        """
+        This takes a register name and returns a register code as per:
+            A = 0b00
+            B = 0b01
+            C = 0b10
+            ...
+            S = 0b111
```
Comment (cannotparse) [inline ToolChain/Assembler/Parser/Parser.py]: to S2

File: ToolChain/Assembler/Parser/Parser.py [hunk 31]
```diff
@@ -517,6 +429,68 @@ def _getInstructionCodeAndFormUsingPossibleCodes(self,
...
+
+            return
+
+        if DATA_ALPHA_INDICATOR in buildInstruction.operationMnemonic:
+            # Treat this as an alphabetic field
+            # The alpha value will be put in the source immediate value as well
+            # We need to rebuild the alpha into a byte string to be read by the VM
+            # We want to preserve spaces in the original string, so we limit the split to the first argument
+
+            textline = line.split(maxsplit=1)
```
Comment (cannotparse) [inline ToolChain/Assembler/Parser/Parser.py]: This is the same as `line[1:]`, which is what was done before.

File: ToolChain/Assembler/Parser/Parser.py [hunk 32]
```diff
@@ -517,6 +429,68 @@ def _getInstructionCodeAndFormUsingPossibleCodes(self,
 
         return instructionCode, instructionForm
 
+    def _parseIndicatorData(self, line, buildInstruction):
+
+        '''
+        Helper method to segregate the evaluateAndExtractInfoFromLine method's operations.
+        This is called if the first argument of a line isn't an instruction. Cases could be:
+            -Global memory reference label
```
Comment (cannotparse) [inline ToolChain/Assembler/Parser/Parser.py]: Kinda inconsistent itemization here.

**Comment by cannotparse** [comment]
Based on #21, I've realized that the code is in fact different. Very suttle. While my opinion on it doesn't change. 

I think we should just start wrapping strings with "". Would make it better. 

This also adds more value to my comment about creating an LR0 based parser for this.

File: ToolChain/Assembler/Parser/Parser.py (hunk 30 above)
Comment (kommisar5150) [reply ToolChain/Assembler/Parser/Parser.py]: Good point! Though @pg-cse suggested I try to keep comments as "future proof" as possible, so it's not impossible for future implementations to have more than just A-S2 registers. I may simply change the wording on this to make it more generic.

File: ToolChain/Assembler/Parser/Parser.py (hunk 8 above)
Comment (kommisar5150) [reply ToolChain/Assembler/Parser/Parser.py]: The idea was to have more "utility" methods at the end, apart from the main functions. Though this could be subjective, perhaps there is a better way to organize the methods? Alphabetical could work as well.

**Comment by cannotparse** [comment]
Here's a simple way to fix my issues with the line splitting 
- Change `_evaluateAndExtractInfoFromLine` to include `raw_line` as a param (and return line to a line segement list)
- Essentially keep your changes but apply the `.split(maxsplit=1)` to `raw_line` then perform your operations on the results

**Comment by pg-cse** [comment]
@cannotparse Good point in https://github.com/CommunicationsSecurityEstablishment/spartacus/pull/20#issuecomment-393563230
It might be time to let go of the old way strings were build and let "" be used as it should have been from the start.

@cannotparse @hopel-cse  @kommisar5150  @psyker156 
Do we all feel the same on this?

**Comment by pg-cse** [comment]
@kommisar5150 Feel free to add commits to this PR. I'd rather keep track of code updates and reasons/recommendations that these updates were mare :)

File: ToolChain/Assembler/Parser/Parser.py [hunk 33]
```diff
@@ -187,18 +186,15 @@ def _parseCodeLine(self, line: str=None):
         possibleCodes = None
 
         # We evaluate the line and extract any bits of relevant info to build the appropriate instruction
-        self._evaluateAndExtractInfoFromLine(line, buildInstruction)
+        self._evaluateAndExtractInfoFromLine(line, buildInstruction, raw_line)
```
Comment (psyker156) [inline ToolChain/Assembler/Parser/Parser.py]: I think raw_line should be renamed to rawLine as all variable in spartacus are camelCase

Comment (pg-cse) [reply ToolChain/Assembler/Parser/Parser.py]: @kommisar5150 can you fix this please?

File: ToolChain/Assembler/Parser/Parser.py [hunk 34]
```diff
@@ -160,184 +175,93 @@ def _parseCodeLine(self, line: list=None):
...
+        self._evaluateAndExtractInfoFromLine(line, buildInstruction, rawLine)
 
         if type(buildInstruction.instructionCode) is not str:
-            # Instruction is real not just memory address or comment
+            # We assume it's an instruction (integer), since memory references or comments would be strings
+            # Parser will verify all possible codes, and generate the appropriate instruction based on its form
             possibleCodes = self._findPossibleCodesForInstruction(buildInstruction)
             instructionCode, instructionForm = self._getInstructionCodeAndFormUsingPossibleCodes(buildInstruction,
-                                                                                                 possibleCodes)
+                                                                                          possibleCodes)
```
Comment (pg-cse) [inline ToolChain/Assembler/Parser/Parser.py]: Parameters should all be aligned.

File: ToolChain/Assembler/Parser/Parser.py [hunk 35]
```diff
@@ -160,184 +175,93 @@ def _parseCodeLine(self, line: list=None):
...
             if instructionForm is None or instructionCode is None:
                 raise ValueError("Assembler general error")
+
```
Comment (pg-cse) [inline ToolChain/Assembler/Parser/Parser.py]: No blank line needed

File: ToolChain/Assembler/Parser/Parser.py [hunk 36]
```diff
@@ -160,184 +175,93 @@ def _parseCodeLine(self, line: list=None):
...
             else:
+                # Code is valid, fetch the instruction's form based on the arguments, then generate the instruction
                 buildInstruction.instructionCode = instructionCode
                 buildInstruction.instructionLength = formDescription[instructionForm]["length"]
                 lineInstruction = self._generateBinaryInstructionFromInstruction(buildInstruction,
                                                                                  formDescription[instructionForm])
         else:
+            # In this case, there were no instructions. Thus, we look at other possibilities
+
```
Comment (pg-cse) [inline ToolChain/Assembler/Parser/Parser.py]: No blank line needed

File: ToolChain/Assembler/Parser/Parser.py [hunk 37]
```diff
@@ -160,184 +175,93 @@ def _parseCodeLine(self, line: list=None):
...
+        This is a helper method. The aim to this is to keep caller code cleaner
+        Method evaluates each part of the line to determine whether we're dealing with an instruction, label, data, etc.
+        We begin by verifying if the first argument has label or data mnemonics (not an instruction).
+        If none are found, we assume it's an instruction. We then proceed to evaluating the arguments.
+        The method uses the Instruction class' source and destination fields to enforce proper instruction format.
+        For example, if the source and destination have already been found, additional arguments means the line is an
+        invalid instruction.
+        :param line:
+        :param buildInstruction:
+        :param rawLine
```
Comment (pg-cse) [inline ToolChain/Assembler/Parser/Parser.py]: Parameters need typing and description information

File: ToolChain/Assembler/Parser/Parser.py [hunk 38]
```diff
@@ -446,20 +346,80 @@ def _evaluateAndExtractInfoFromLine(self, line, buildInstruction):
                     part = part[1:]
                 if not foundSource:
                     foundSource = True
+
                     # Verify this reference doesn't have additonal ':', hotfix for issue 18
                     if part.count(':') > 0:
                         raise ValueError("Syntax error, memory reference has too many {}".format(MEMORY_REFERENCE_INDICATORS))
-                    buildInstruction.sourceImmediate = MEMORY_REFERENCE_INDICATORS+ part.upper() + MEMORY_REFERENCE_INDICATORS
+                    buildInstruction.sourceImmediate = MEMORY_REFERENCE_INDICATORS + part.upper() + \
+                                                       MEMORY_REFERENCE_INDICATORS
```
Comment (pg-cse) [inline ToolChain/Assembler/Parser/Parser.py]: In this case, for readability, I would put part.upper() on a different line as well.

File: ToolChain/Assembler/Parser/Parser.py [hunk 39]
```diff
@@ -446,20 +346,80 @@ def _evaluateAndExtractInfoFromLine(self, line, buildInstruction):
...
                 else:
                     if foundDestination:
                         raise ValueError("Invalid operation format")
                     foundDestination = True
+
                     # Verify this label doesn't have additonal ':', hotfix for issue 18
                     if part.count(':') > 0:
                         raise ValueError("Syntax error, memory reference has too many {}".format(MEMORY_REFERENCE_INDICATORS))
-                    buildInstruction.destinationImmediate = MEMORY_REFERENCE_INDICATORS + part.upper() + MEMORY_REFERENCE_INDICATORS
+                    buildInstruction.destinationImmediate = MEMORY_REFERENCE_INDICATORS + part.upper() + \
```
Comment (pg-cse) [inline ToolChain/Assembler/Parser/Parser.py]: In this case, for readability, I would put part.upper() on a different line as well.

File: ToolChain/Assembler/Parser/Parser.py [hunk 40]
```diff
@@ -446,20 +346,80 @@ def _evaluateAndExtractInfoFromLine(self, line, buildInstruction):
...
+        Helper method to segregate the evaluateAndExtractInfoFromLine method's operations.
+        This is called if the first argument of a line isn't an instruction. Cases could be:
+        -Global memory reference label
+        -Local memory reference label
+        -Comment
+        -.dataAlpha, .dataNumeric, or .dataMemref
+        :param line:
+        :param buildInstruction:
+        :param rawLine
+        :return:
```
Comment (pg-cse) [inline ToolChain/Assembler/Parser/Parser.py]: Parameters and return value need typing and description information.

File: ToolChain/Assembler/Parser/Parser.py [hunk 41]
```diff
@@ -446,20 +346,80 @@ def _evaluateAndExtractInfoFromLine(self, line, buildInstruction):
...
+            buildInstruction.instructionLength = 4
+            buildInstruction.instructionCode = line[0].upper()
+            return
+
+        if DATA_ALPHA_INDICATOR in buildInstruction.operationMnemonic:
+            # Treat this as an alphabetic field
+            # The alpha value will be put in the source immediate value as well
+            # We need to rebuild the alpha into a byte string to be read by the VM
+            # We want to preserve spaces in the original string, so we limit the split to the first argument
+
```
Comment (pg-cse) [inline ToolChain/Assembler/Parser/Parser.py]: No new line so it's coherent with the other blocks.

File: ToolChain/Assembler/Parser/Parser.py [hunk 42]
```diff
@@ -446,20 +346,80 @@ def _evaluateAndExtractInfoFromLine(self, line, buildInstruction):
...
+            alpha = alpha[0:-1]  # Remove trailing space
+            alpha += b"\x00"  # Strings are null terminated
+            buildInstruction.sourceImmediate = alpha
+            buildInstruction.instructionLength = len(alpha)  # Length of the instruction is simply length of string
+            buildInstruction.instructionCode = line[0].upper()
+            return
+
+        if DATA_MEMORY_REFERENCE in buildInstruction.operationMnemonic:
+            # This is a memory reference used as a constant that needs to be linked later in the process.
+            if line[1][0] == ":":
```
Comment (pg-cse) [inline ToolChain/Assembler/Parser/Parser.py]: ":" should be replaced with MEMORY_REFERENCE_INDICATORS

File: ToolChain/Assembler/Parser/Parser.py [hunk 43]
```diff
@@ -446,20 +346,80 @@ def _evaluateAndExtractInfoFromLine(self, line, buildInstruction):
...
+                line[1] = line[1][1:]
+
+            buildInstruction.sourceImmediate = (":" + line[1] + ":").upper().encode("utf-8")
```
Comment (pg-cse) [inline ToolChain/Assembler/Parser/Parser.py]: ":" should be replaced with MEMORY_REFERENCE_INDICATORS

File: ToolChain/Assembler/Parser/Parser.py [hunk 44]
```diff
@@ -585,3 +548,131 @@ def _generateBinaryInstructionFromInstruction(self, instruction: Instruction=Non
...
+        This will translate an immediate value in a way that can be understood by the architecture.
+        :param textImmediate: str, an immediate value to be translated
+        :return: int, an immediate that can be worked on
+        """
+
+        immediate = None
+        isNegative = False
+        textImmediate = textImmediate.lower()  # Needed in case of 0XFF instead of 0xFF
+
+        if textImmediate[0] == "-":
```
Comment (pg-cse) [inline ToolChain/Assembler/Parser/Parser.py]: Maybe we want to add a new constant to 
spartacus/ToolChain/Assembler/Constants.py

File: ToolChain/Assembler/Parser/Parser.py [hunk 45]
```diff
@@ -585,3 +548,131 @@ def _generateBinaryInstructionFromInstruction(self, instruction: Instruction=Non
...
+            isNegative = True
+            textImmediate = textImmediate[1:]
+
+        if len(textImmediate) > 2 and textImmediate[0:2] == "0b":
```
Comment (pg-cse) [inline ToolChain/Assembler/Parser/Parser.py]: Maybe we want to add a new constant to 
spartacus/ToolChain/Assembler/Constants.py

File: ToolChain/Assembler/Parser/Parser.py [hunk 46]
```diff
@@ -585,3 +548,131 @@ def _generateBinaryInstructionFromInstruction(self, instruction: Instruction=Non
...
+            # Indicates binary immediate
+            baseToUse = 2
+            textImmediate = textImmediate[2:]
+        elif len(textImmediate) > 2 and textImmediate[0:2] == "0x":
```
Comment (pg-cse) [inline ToolChain/Assembler/Parser/Parser.py]: Maybe we want to add a new constant to 
spartacus/ToolChain/Assembler/Constants.py

File: ToolChain/Assembler/Parser/Parser.py [hunk 47]
```diff
@@ -446,20 +346,80 @@ def _evaluateAndExtractInfoFromLine(self, line, buildInstruction):
...
+                    buildInstruction.sourceImmediate = MEMORY_REFERENCE_INDICATORS + part.upper() + \
+                                                       MEMORY_REFERENCE_INDICATORS
+                    
                 else:
                     if foundDestination:
                         raise ValueError("Invalid operation format")
                     foundDestination = True
+
                     # Verify this label doesn't have additonal ':', hotfix for issue 18
                     if part.count(':') > 0:
```
Comment (pg-cse) [inline ToolChain/Assembler/Parser/Parser.py]: I know this is not new code but this comment applies. If we are updating the file I think it should be cleaned up
":" should be replaced with MEMORY_REFERENCE_INDICATORS

File: ToolChain/Assembler/Parser/Parser.py [hunk 48]
```diff
@@ -446,20 +346,80 @@ def _evaluateAndExtractInfoFromLine(self, line, buildInstruction):
                     part = part[1:]
                 if not foundSource:
                     foundSource = True
+
                     # Verify this reference doesn't have additonal ':', hotfix for issue 18
                     if part.count(':') > 0:
```
Comment (pg-cse) [inline ToolChain/Assembler/Parser/Parser.py]: I know this is not new code but this comment applies. If we are updating the file I think it should be cleaned up
":" should be replaced with MEMORY_REFERENCE_INDICATORS

File: ToolChain/Assembler/Parser/Parser.py (hunk 45 above)
Comment (cannotparse) [reply ToolChain/Assembler/Parser/Parser.py]: For fun, I went and looked for how to auto interpret the results. TIL: `int(intstring,0)` will auto interpret the value and find the appropriate base if '0b' or '0x' are present. It isn't a magic bullet though. 

The result would look like this:
```python3 
try:
   immediate = int(textImmediate, 0)
except ValueError as e:
   raise(ValueError("Invalid Immediate format provided. Try placing {}, {}, or {} infront of the value. ").format('0x', '0b', '0o'))
``` 
I don't know how you feel about this change.

**Comment by pg-cse** [comment]
Good fix for https://github.com/CommunicationsSecurityEstablishment/spartacus/issues/21
